src/Poly.jl

```diff
@@ -2148,29 +2148,94 @@ end
 ###############################################################################
 
 @doc raw"""
-    evaluate(a::PolyRingElem, b::T) where T <: RingElement
+    evaluate(a::PolyRingElem, b)
 
 Evaluate the polynomial expression $a$ at the value $b$ and return the result.
 """
-function evaluate(a::PolyRingElem, b::T) where T <: RingElement
+function evaluate(a::PolyRingElem, b)
    i = length(a)
    R = base_ring(a)
    S = parent(b)
    if i == 0
       return zero(R) + zero(S)
    end
    if i > 25
-      return subst(a, b)
+      return evaluate_brent_kung(a, b)
+   end
+   return evaluate_horner(a, b)
+end
+
+function evaluate_horner(a::PolyRingElem, b)
+   i = length(a)
+   R = base_ring(a)
+   S = parent(b)
+   if i == 0
+      return zero(R) + zero(S)
    end
    z = coeff(a, i - 1) * one(S)
    while i > 1
       i -= 1
-      z = coeff(a, i - 1) + z*b
-      parent(z) # To work around a bug in julia
+      z = z * b
+      # TODO: change to mul!(z, z, b) once Nemo & Hecke require AA 0.45
+      z = add!(z, z, coeff(a, i - 1))
    end
    return z
 end
 
+# uses the algorithm of "A Practical Univariate Polynomial Composition Algorithm"
+# this is asymptotically faster, if scalar multiplications are cheaper than
+# base ring multiplications
```

Comment:
I realize this is old stuff you just moved, but still out of curiosity: what does this mean? A scalar multiplication means multiplying a polynomial by a coefficient from the base ring of the polynomial. Surly that involves multiple base ring multiplications. How can this be a faster than a single base ring multiplication?

So probably my guess at an interpretation is simply wrong, but what is really meant then?

Reply:
Assume you have a polynomial $f \in K[X]$ and evaluate at an element $a$ of a $K$-algebra $A$. Multiplication of elements of $K$ with $A$ is the "scalar multiplication" and "base ring multiplication" refers to multiplication in $A$. This algorithm tries to do less multiplications in $A$ at the cost of more scalar multiplications. Typical example would be evaluation of a rational polynomial at a rational matrix.

Reply:
thanks that explains it, but I would not have been (in fact: was not) able to infer this from the current comment. Perhaps some or all of what you wrote could be copied into the comment?

Reply:
I rewrote the comment